src/beancount.pest

```diff
@@ -42,61 +45,116 @@ commodity = @{ ASCII_ALPHA_UPPER ~ commodity_trailing{1, 23} }
 //   SELECT account, sum(position) WHERE ‘trip-france-2014’ in tags"
-query = { date ~ WHITE_SPACE ~ "query" ~ WHITE_SPACE ~ quoted_str ~ WHITE_SPACE ~ quoted_str ~ eol ~ key_value_list? }
+query = { date ~ ws+ ~ "query" ~ ws+ ~ quoted_str ~ ws+ ~ quoted_str ~ eol ~ key_value_list? }
+
+//// Transaction directive
 
 // 2014-05-05 txn "Cafe Mogador" "Lamb tagine with wine"
 //     Liabilities:CreditCard:CapitalOne         -37.45 USD
 //     Expenses:Restaurant
-txn_indicator  = { "txn" | "*" | "!" }
+transaction = { date ~ ws+ ~ txn_flag ~ ws+ ~ txn_strings ~ (ws+ ~ tags_links)? ~ eol ~ posting_or_kv_list? }
+txn_flag = { flag_okay | flag_warning | flag_padding | flag_summarize | flag_transfer | flag_conversions | flag_unrealized | flag_returns | flag_merging | flag_hash }
+// Transactions that have been checked.
+flag_okay = { "*" | "txn" }
+// Mark by the user as something to be looked at later on.
+flag_warning = { "!" }
+// Transactions created from padding directives.
+flag_padding = { "P" }
+// Transactions created due to summarization.
+flag_summarize = { "S" }
+// Transactions created due to balance transfers.
+flag_transfer = { "T" }
+// Transactions created to account for price conversions.
+flag_conversions = { "C" }
+// Transactions created due to unrealized gains.
+flag_unrealized = { "U" }
+// Transactions that were internalized by returns algorithm.
+flag_returns = { "R" }
+// A flag to mark postings merging together legs for average cost.
+flag_merging = { "M" }
+// TODO: Figure out what this flag means and document accordingly.  It is supported in the original grammar, but I'm not sure what it is.
+// https://bitbucket.org/blais/beancount/src/default/beancount/parser/grammar.y?fileviewer=file-view-default#grammar.y-274
+// https://github.com/twilco/beancount/issues/10
+flag_hash = { "#" }
 // TODO: Consider adding support for deprecated pipe syntax.  https://bitbucket.org/blais/beancount/src/default/beancount/parser/grammar.y?fileviewer=file-view-default#grammar.y-341
 // https://github.com/twilco/beancount/issues/2
-txn_strings = { quoted_str ~ (WHITE_SPACE ~ txn_strings)? }
-transaction = { date ~ WHITE_SPACE ~ txn_indicator ~ WHITE_SPACE ~ txn_strings }
+txn_strings = { quoted_str ~ (ws+ ~ quoted_str)? }
+posting = {
+    //   ! Assets:BofA:Checking 1234.32 USD {{502.12 # 9.95 USD, 2018-01-01}} @@ 173.12 US
+	indent ~ (txn_flag ~ ws+)? ~ account ~ ws+ ~ incomplete_amount ~ (ws+ ~ cost_spec)? ~ (ws+ ~ ("@@" | "@") ~ ws+ ~ price_annotation)? ~ eol
+    |
+    //   S Assets:BofA:Checking
+    indent ~ (txn_flag ~ ws+)? ~ account ~ eol
+}
+posting_or_kv_list = {
+  (key_value_line | posting | (indent ~ tags_links ~ NEWLINE) | (indented_comment_line))*
+}
+price_annotation = { incomplete_amount }
+// 189.10 USD | USD | 152.41
+incomplete_amount = { (num_expr ~ ws+ ~ commodity) | commodity | num_expr }
+// {{502.12 # 9.95 USD, 2018-01-01}}
+cost_spec = { ("{{" | "{") ~ cost_comp_list ~ ("}}" | "}") }
```

Comment:
This matches unbalanced braces: `{{ ... }`. Should either be split into two rules or use `PUSH` and `POP`.

Reply:
Ahh, great catch.  Fixed.